torch/nested/_internal/sdpa.py

```diff
@@ -0,0 +1,723 @@
+        _check_for_seq_len_0_nested,
+    )
+    for constraint in constraints:
+        if not constraint(params, debug):
+            return False
+    return True
+
+
+def _can_use_math_sdpa_jagged(params: SDPAParams, debug=False) -> bool:
+    if (
+        not params.query.is_contiguous()
+        or not params.key.is_contiguous()
+        or not params.value.is_contiguous()
+    ):
+        if debug:
+            log.warning("If inputs are nested tensors they must be contiguous.")
+        return False
+    if params.is_causal:
+        if debug:
+            log.warning(
+                "Nested tensors for query / key are not supported when is_causal=True."
+            )
+        return False
+    return True
+
+
+def _select_sdp_backend(query, key, value, attn_mask, dropout, is_causal):
+    if (
+        not flash_sdp_enabled()
+        and not mem_efficient_sdp_enabled()
+        and not math_sdp_enabled()
+    ):
+        return SDPBackend.ERROR
+
+    ordering = (
+        SDPBackend.FLASH_ATTENTION,
+        SDPBackend.EFFICIENT_ATTENTION,
+        SDPBackend.MATH,
+    )
+
+    params = SDPAParams(query, key, value, attn_mask, dropout, is_causal)
+
+    for backend in ordering:
+        if backend == SDPBackend.FLASH_ATTENTION:
+            if can_use_flash_attention(params) and _can_use_flash_sdpa_jagged(params):
+                return SDPBackend.FLASH_ATTENTION
+        if backend == SDPBackend.EFFICIENT_ATTENTION:
+            if can_use_efficient_attention(params) and _can_use_efficient_sdpa_jagged(
+                params
+            ):
+                return SDPBackend.EFFICIENT_ATTENTION
+        if backend == SDPBackend.EFFICIENT_ATTENTION:
+            if math_sdp_enabled() and _can_use_math_sdpa_jagged(params):
+                return SDPBackend.MATH
+
+    log.warning("Memory efficient kernel not used because:")
+    can_use_efficient_attention(params, debug=True)
+    _can_use_efficient_sdpa_jagged(params, debug=True)
+    log.warning("Flash attention kernel not used because:")
+    can_use_flash_attention(params, debug=True)
+    _can_use_flash_sdpa_jagged(params, debug=True)
+    raise ValueError("No available kernel. Aborting execution.")
+
+
+def _cumulative_and_max_seq_len_nnz(qkv):
+    """
+    This function is used to calculate two pieces of metadata that are needed
+    for use with flash-attention and efficient_attention kernels. They are the
+    cumulative sequence_length over a batch of sequences and the maximum
+    sequence length.
+
+    @return A tuple of cumulative sequence lengths and the maximum sequence
+    length, and the last element in the cumulative_sequence_lengths
+    """
+    if not qkv.is_nested:
+        raise ValueError("QKV must be nested for flash cumulative_seq_len calculation.")
+
+    if qkv.lengths() is None:
+        cumulative_seqlen = qkv.offsets().to(dtype=torch.int32, device=qkv.device)
+        max_seqlen = torch.max(qkv.offsets().diff()).item()
```

Comment:
lets add a some TODOS/ notes around the performance for when we need to sync and call .item()